### ISSUE TITLE
GraphQL - Create GraphQL Service Abstraction 

### ISSUE THREAD

**Issue by erikschlegel** [CatalystCode/project-fortis-spark#17]
This is meant to be a thin wrapper for integrating the newly modified graphql service resolver response to translate to the data structure that the V1 fortis dashboard expects. 

*what complete means*
Integrate this abstraction layer into fortis-interfaces services.js file.

**Comment by c-w** [comment]
Blocked on schema being finalized and test Cassandra DB being set up.

**Comment by c-w** [comment]
My understanding of the situation is the following:

1) The [project-fortis-interfaces](https://github.com/CatalystCode/project-fortis-interfaces) frontend posts graphql queries to the [project-fortis-services](https://github.com/CatalystCode/project-fortis-services) backend ([code](https://github.com/CatalystCode/project-fortis-interfaces/blob/347c4464e5aa0125df30feadb25035dee79c431b/src/services/services.js#L171-L179)).

2) The fortis-services backend defines some [resolvers](https://github.com/CatalystCode/project-fortis-services/tree/35424cc31f9685d3eecb87bda873e4e6c7de8b0b/src/resolvers) for the queries.

3) For minimal change, we can just re-implement all the fortis-services resolvers in terms of the new Cassandra backend. This enables us to not have to touch the frontend at all.

**Comment by c-w** [comment]
Node driver for Cassandra: [datastax/nodejs-driver](https://github.com/datastax/nodejs-driver#prepare-your-queries)

**Comment by c-w** [comment]
Here's a list of all the currently implemented GraphQL endpoints in Fortis:

## Mutations ##

### EdgesSchema.js ###

```gql
removeKeywords
addKeywords
removeLocations
saveLocations
```

### MessageSchema.js ###

```gql
publishEvents
```

### SettingsSchema.js ###

```gql
removeFacebookPages
modifyFacebookPages
createOrReplaceSite
modifyTwitterAccounts
removeTwitterAccounts
modifyTrustedTwitterAccounts
removeTrustedTwitterAccounts
modifyBlacklist
removeBlacklist
```

## Queries ##

### EdgesSchema.js ###

```gql
locations
terms
popularLocations
timeSeries
topSources
```

### FactSchema.js ###

```gql
list
get
```

### MessageSchema.js ###

```gql
byBbox
byEdges
event
translate
translateWords
```

### SettingsSchema.js ###

```gql
sites
twitterAccounts
trustedTwitterAccounts
facebookPages
facebookAnalytics
termBlacklist
```

### TilesSchema.js ###

```gql
fetchTilesByBBox
fetchPlacesByBBox
fetchTilesByLocations
fetchEdgesByLocations
fetchEdgesByBBox
```

### TopLocationsSchema.js ###

```gql
popularLocations
```

**Comment by c-w** [comment]
TopLocationsSchema looks unused.

**Comment by c-w** [comment]
[FactSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/FactSchema.js) can be implemented in terms of `fortis.events`.

[SettingsSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/SettingsSchema.js) can be implemented in terms of `fortis.sitesettings`.

I don't see a correspondence between [EdgesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/EdgesSchema.js), [TilesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/TilesSchema.js), [MessagesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/MessageSchema.js) and the new Cassandra schema. @erikschlegel could you help bridge my understanding gap here?

**Comment by erikschlegel** [comment]
Correct, `TopLocationsSchema.js` is not used. @Smarker started the work with the Cassandra driver manager which we should use as a starting point for integrating with our Cassandra Contact Points.  

Only other thing item worth mentioning is that subscriptions will eventually be added to this abstraction layer once it's ready, as a separate issue. 

Agree with all statements mentioned above.

**Comment by c-w** [comment]
@erikschlegel Could you please explain how the new proposed Cassandra schema ([reference](https://github.com/CatalystCode/project-fortis-spark/issues/14#issuecomment-310725868)) maps to the schemas defined by [EdgesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/EdgesSchema.js), [TilesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/TilesSchema.js) and [MessagesSchema.js](https://github.com/CatalystCode/project-fortis-services/blob/1c2e6ad6939d0c3e5eed2e6e9622460ae714103c/src/schemas/MessageSchema.js)?

It seems as though the Cassandra schema and the GraphQL schemas have vastly different vocabularies so it would be great if you could help to translate between the two worlds since you're the only one on the team familiar with the Fortis-v1 terminology.

**Comment by c-w** [comment]
Went through all the tables with Erik which was very productive. In general, the mapping between GraphQL and Cassandra gets much clearer if we think that "edge" is the same thing as "topic".

**Comment by c-w** [comment]
Here is the correspondence between GraphQL and Cassandra:

## Mutations ##

### ~~MessageSchema.js~~ ###

~~publishEvents --> EventHub~~

### SettingsSchema.js ###

{modify,remove}Blacklist --> fortis.blacklist
{modify,remove}TrustedTwitterAccounts --> fortis.trustedsources
{modify,remove}TwitterAccounts --> fortis.streams
{modify,remove}FacebookPages --> fortis.trustedsources
createOrReplaceSite --> fortis.sitesettings

### ~~EdgesSchema.js~~ ###

~~addKeywords --> fortis.watchlist~~
~~removeKeywords --> fortis.watchlist~~
~~saveLocations --> spark~~
~~removeLocations --> spark~~

## Queries ##

### ~~TilesSchema.js~~ ###

~~fetchTilesByBBox --> fortis.computedtiles~~
~~fetchTilesByLocations --> fortis.computedtiles~~
~~fetchPlacesByBBox --> feature service~~
~~fetchEdgesByBBox --> fortis.computedtiles // edge == topic~~
~~fetchEdgesByLocations --> fortis.computedplaces // edge == topic~~

### SettingsSchema.js ###

sites --> fortis.sitesettings
trustedTwitterAccounts --> fortis.trustedsources
twitterAccounts --> fortis.streams
facebookPages --> fortis.trustedsources
~~facebookAnalytics --> facebook api~~
termBlacklist --> fortis.blacklist

### ~~MessageSchema.js~~ ###

~~byBbox --> fetch wofids from feature service and then look at fortis.events~~
~~byEdges --> fortis.events // edge == keyword~~
~~event --> fortis.events~~
~~translate --> current implementation~~
~~translateWords --> current implementation~~

### ~~FactSchema.js~~ ###

~~list --> fortis.events // fact == event that is not twitter or facebook~~
~~get --> fortis.events // fact == event that is not twitter or facebook~~

### EdgesSchema.js ###

~~terms --> fortis.events // detectedkeywords~~
~~locations --> get geofence from fortis.sitesettings and then query feature service~~
popularLocations --> get geofence from fortis.sitesettings and then fortis.computedtiles
timeSeries --> fortis.computedtiles
topSources --> fortis.computedtiles

**Comment by c-w** [comment]
All the follow-up tasks captured. Now it's just a question of implementing each of the resolvers. This will be quite a big effort so should be shared across the team.

**Comment by c-w** [comment]
Split the task into [project-fortis#49](https://github.com/CatalystCode/project-fortis/issues/49), [project-fortis#50](https://github.com/CatalystCode/project-fortis/issues/50), [project-fortis#51](https://github.com/CatalystCode/project-fortis/issues/51) and [project-fortis#52](https://github.com/CatalystCode/project-fortis/issues/52).
